### ISSUE TITLE
Loading text with BOM encoding

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#5268]
## tl;dr

A community zync user mentioned the following in a Zui context, though it seems addressing it would likely have to start at the Zed layer since that's what handles data loading. In their own words:

>I noticed Zui does not properly ingest JSON files which have been saved using UTF8BOM encoding, while UTF8NOBOM works well. I was also asking myself whether it would useful to be able to specify the file encoding in Zui, à-la `iconv`.

## Details

Repro is with Zui Insiders [1.17.1-insiders.20](https://github.com/brimdata/zui-insiders/releases/tag/v1.17.1-insiders.20) which uses Zed commit 556f586.

I was not personally familiar with [byte-order mark (BOM)](https://en.wikipedia.org/wiki/Byte_order_mark) but gave myself a quick crash course so I could repro the user's experience.

This first video shows an example in the Windows context. Notepad's default selection to **Save As** in "UTF-8" creates a text file that Zui loads without complaint. However, if the "UTF-8 with BOM" option is selected, indeed the text format is no longer recognized by Zed's auto-detect as a format that can be read.

https://github.com/user-attachments/assets/dabf3a7e-237b-4f6c-819c-597a70c0b07f

That same [page on Wikipedia](https://en.wikipedia.org/wiki/Byte_order_mark) also mentions that Google Docs adds the BOM when selecting the "Plain Text" download format, and indeed, this second video shows this causing the same error message we just saw with the file saved from Notepad.

https://github.com/user-attachments/assets/c1207904-9656-4071-98e5-cb109305a0a8

I'm not certain if it's feasible for Zed's auto-detect to recognize and react to the BOM without disturbing the ability to auto-detect the other formats supported by Zed. If not, I suppose we could add an explicit reader options indicating to expect the BOM and read in the specified format, per the user's `iconv` comment.

I don't know if it would do the trick, but I did some web searches and found https://pkg.go.dev/github.com/dimchansky/utfbom which it sounds like it may be what's needed:

> The package utfbom implements the detection of the BOM (Unicode Byte Order Mark) and removing as necessary. It can also return the encoding detected by the BOM.

If the returned encoding is accurate, then perhaps this is the info that would be needed to automatically react to the BOM when it's present and read in the encoded format.

**Comment by philrz** [comment]
The same user whose experience originally inspired the opening of this issue circled back with another BOM encounter, this time involving CSV exported from Excel. The attached video captures my repro with super commit 15d2ef0 and the [with-bom.csv](https://github.com/user-attachments/files/26490943/with-bom.csv) file created in the video.

https://github.com/user-attachments/assets/b935369a-9183-4828-b967-a58a9e11549c

As the video shows, all it takes to get a CSV with the leading BOM is to select the **Download as CSV UTF-8** option, which is the top CSV export option shown in cloud Excel's drop-down menu. It seems that when rendered to the terminal as text with `cat` or even `super`, the leading BOM is rendered as a zero-width, invisible character, so the user may not even be aware it's present.

```
$ cat with-bom.csv 
a,b
foo,bar

$ super -version && super with-bom.csv 
Version: v0.3.0-28-g15d2ef03a
{"﻿a":"foo",b:"bar"}
```

However, once we try to access the first field by name, we trigger the problem: It seems likely the leading bytes that make up the BOM interfere with our ability to access the first field name in the CSV header line.

```
$ super -c 'cut b' with-bom.csv 
{b:"bar"}

$ super -c 'cut a' with-bom.csv 
no such field "a" at line 1, column 5:
cut a
    ~
no such field "a" at line 1, column 5:
cut a
    ~
```

The [BOM wikipedia page](https://en.wikipedia.org/wiki/Byte_order_mark) points out:

> The BOM is, simply, the Unicode codepoint `U+FEFF` (ZERO WIDTH NO-BREAK SPACE), encoded in the current encoding.

So with knowledge of this and a little hacking, a user could do the following, which indicates the BOM bytes are being treated as part of the first field name.

```
$ super -c 'head 1 | values {a:this["\uFEFFa"]}' with-bom.csv 
{a:"foo"}
```

Seen as raw bytes, the UTF-8 BOM looks like:

```
$ head -c 3 with-bom.csv | xxd
00000000: efbb bf                                  ...
```

Consensus from AI systems aligned with my suspicion that many tools deal with this by putting BOM-stripping logic at the entry point of text-based readers, but not attempting the same with binary readers since there's a chance the byte sequence `0xEF 0xBB 0xBF` could appear at the start of a binary stream for other legitimate reasons having nothing to do with BOM.

Therefore a brute force way to strip the BOM when it's known to be present might be:

```
$ tail -c +4 with-bom.csv > without-bom.csv

$ super -c 'cut a' without-bom.csv 
{a:"foo"}
```

Though a more sophisticated approach could look for and only skip the specific BOM bytes.
